### ISSUE TITLE
"Dangling" means multiple things

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/unsafe-code-guidelines#478]
In the C++ world, typically, a "dangling" pointer is one to memory that was previously allocated and subsequently deallocated. Let's call these freed-dangling. These are just bad, period.

In Rust, APIs like [`NonNull::dangling()`](https://doc.rust-lang.org/stable/std/ptr/struct.NonNull.html#method.dangling)` produce pointers that were not ever allocated, primarily constructed to be "aligned but not null".. Let's call these align-dangling. It's good practice to use align-dangling pointers when working with collections or ZSTs.

We tend to use these meanings interchangeably. They are _largely_ the same when it comes to the validity of pointer accesses, [except around zero-sized types](https://doc.rust-lang.org/std/ptr/index.html#safety): zero-sized reads are not valid when reading from freed memory, unless the pointer was obtained as a direct cast from an integer literal (presumably these get normalized to align-dangling pointers).

This can lead to some confusion around terminology, for example with [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=445fb0d4b0a80b0628f904320ff23d60) (credit @kupiakos), where we have both a freed-dangling and an align-dangling `NonNull::dangling()` pointer, and miri complains about the former, but says "out-of-bounds pointer use: alloc921 has been freed, so this pointer is dangling", implicitly calling it "dangling".

We should potentially pin down what meaning of "dangling" we intend everywhere, use it consistently, and update the docs of `NonNull::dangling()` to reference ZST validity.

**Comment by RalfJung** [comment]
I consider there to be three kinds of "dangling" pointers:
- pointers with no provenance whatsoever -- that's what you call "align-dangling", but it has nothing to do with alignment; this is what `ptr::invalid` (maybe [to be renamed](https://github.com/rust-lang/rust/pull/117658)) creates
- pointers with provenance of some allocation that has been freed -- that's what you call "freed-dangling"
- pointers with provenance of some allocation that is still live but the pointer is out-of-bounds of that allocation

I agree it'd be good to clean up the terminology here. Relatedly, the term "dereferenceable" also [needs clarification](https://rust-lang.zulipchat.com/#narrow/stream/136281-t-opsem/topic/Do.20we.20need.20a.20new.20name.20for.20.22dereferenceable.22.3F).

**Comment by Manishearth** [comment]
> but it has nothing to do with alignment so I'd rather avoid that term

Yeah it's typically just _used_ in a case where you want something aligned. I don't have a better term for it, page0-dangling perhaps.

**Comment by RalfJung** [comment]
They can point to literally any address though, `ptr::invalid` takes a usize. I sometimes call them "(raw) integer pointers", but that doesn't really fit the pattern here.

**Comment by digama0** [comment]
> They are _largely_ the same when it comes to the validity of pointer accesses, [except around zero-sized types](https://doc.rust-lang.org/std/ptr/index.html#safety): zero-sized reads are not valid when reading from freed memory, unless the pointer was obtained as a direct cast from an integer literal (presumably these get normalized to align-dangling pointers).

I believe this is no longer true as of the [recent FCP](https://github.com/rust-lang/unsafe-code-guidelines/issues/472): zero sized reads from any pointer are now valid, so AFAIK freed-dangling pointers are indistinguishable from align-dangling (which I would call no-provenance / null-provenance pointers). The linked documentation is out of date.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/issues/117945 is tracking implementing that FCP.
